### PR TITLE
Disable garbage collection during logging

### PR DESCRIPTION
**Pull request by richtja**
In some circumstances, avocado instrumented runner can hang indefinitely and block other test to execute. This is caused by a very specific bug in `multiprocessing.Queue` which can cause deadlock when you're manipulating with queue during garbage collection. This issue was investigated and well described by @pevogam in
https://github.com/avocado-framework/avocado/issues/4935#issuecomment-938069202 for long story short, this is the process which causes deadlock:

   1. The program was calling q.put(2).
   2. This involves acquiring a lock.
   3. Half-way through the function call, garbage collection happens.
   4. Garbage collection calls Object.__del__.
   5. Object.__del__ calls q.put(1).
   6. q.put(1) tries to acquire the lock… but the lock is already held, so it waits.

Because instrumented runner is using multiprocessing.Queue for storing logs, this is quite common scenario if any object does logging in `__del__` method.

This commit disables garbage collection for the process when the logs are stored to `multiprocessing.Queue`. Such change prevents this kind of deadlock.

Reference: #4935

### CODE REVIEW

**Comment by pevogam** [comment]
@richtja This will require some more advanced testing so perhaps I can look into it next week. It is good if we take our time with this since it took quite a while to fix it and a potential fix should be evaluated properly. On my side I also have to restore some behaviors that can only bring up the issue again randomly with increased probability in 10+ hour test runs and I hope I can recreate the issue first before needing long enough testing and usage of this pull request (more days/weeks of testing and usage increase the probability the issue has been fixed) to cross this off on our side. It could be for instance, that there are other locations that will need the GC disabled too, we will see.

In short, the fix seems very reasonable at first look but it will need good care in testing and ruling out.

**Review by clebergnu — APPROVED** [review]
LGTM, thanks!  And also thanks @pevogam for the time and energy spent in the investigation of this issue!

**Comment by clebergnu** [comment]
> @richtja This will require some more advanced testing so perhaps I can look into it next week. It is good if we take our time with this since it took quite a while to fix it and a potential fix should be evaluated properly. On my side I also have to restore some behaviors that can only bring up the issue again randomly with increased probability in 10+ hour test runs and I hope I can recreate the issue first before needing long enough testing and usage of this pull request (more days/weeks of testing and usage increase the probability the issue has been fixed) to cross this off on our side. It could be for instance, that there are other locations that will need the GC disabled too, we will see.
> 
> In short, the fix seems very reasonable at first look but it will need good care in testing and ruling out.

Hi @pevogam, this seems reasonable to me as well.  I'm OK with merging it and testing it on "on the wild".  Are you OK with that too?

**Comment by pevogam** [comment]
> > @richtja This will require some more advanced testing so perhaps I can look into it next week. It is good if we take our time with this since it took quite a while to fix it and a potential fix should be evaluated properly. On my side I also have to restore some behaviors that can only bring up the issue again randomly with increased probability in 10+ hour test runs and I hope I can recreate the issue first before needing long enough testing and usage of this pull request (more days/weeks of testing and usage increase the probability the issue has been fixed) to cross this off on our side. It could be for instance, that there are other locations that will need the GC disabled too, we will see.
> > In short, the fix seems very reasonable at first look but it will need good care in testing and ruling out.
> 
> Hi @pevogam, this seems reasonable to me as well. I'm OK with merging it and testing it on "on the wild". Are you OK with that too?

I would prefer if it waits for its time, the corresponding issue has been waiting for the pull request for months if not years so I believe the pull request could wait for some days at least :o)

**Comment by clebergnu** [comment]
> > > @richtja This will require some more advanced testing so perhaps I can look into it next week. It is good if we take our time with this since it took quite a while to fix it and a potential fix should be evaluated properly. On my side I also have to restore some behaviors that can only bring up the issue again randomly with increased probability in 10+ hour test runs and I hope I can recreate the issue first before needing long enough testing and usage of this pull request (more days/weeks of testing and usage increase the probability the issue has been fixed) to cross this off on our side. It could be for instance, that there are other locations that will need the GC disabled too, we will see.
> > > In short, the fix seems very reasonable at first look but it will need good care in testing and ruling out.
> > 
> > 
> > Hi @pevogam, this seems reasonable to me as well. I'm OK with merging it and testing it on "on the wild". Are you OK with that too?
> 
> I would prefer if it waits for its time, the corresponding issue has been waiting for the pull request for months if not years so I believe the pull request could wait for some days at least :o) This is at least for the most initial bulk testing phase of course, the longer term ("weeks" above) could also do this port-merge since it will block the pull request for too long while the uncertainty reduces below some threshold.

Fine with me!

**Comment by richtja** [comment]
> I would prefer if it waits for its time, the corresponding issue has been waiting for the pull request for months if not years so I believe the pull request could wait for some days at least :o) This is at least for the most initial bulk testing phase of course, the longer term ("weeks" above) could also do this port-merge since it will block the pull request for too long while the uncertainty reduces below some threshold.

I agree, we can wait to make sure that this has been fixed.

**Review by pevogam — APPROVED** [review]
Alright, all tests passed and the changes are very compact for such an issue which is good. Great idea @richtja on how to go around this!

**Comment by richtja** [comment]
thank you @pevogam for testing. We are ready to merge it.
